### PR TITLE
impl Component for World

### PR DESCRIPTION
**Pull request by Xaeroxe**
This PR permits recursive worlds.

### CODE REVIEW

**Comment by torkleyy** [comment]
My response refers to what @Xaeroxe [wrote yesterday](https://gitter.im/amethyst/engine?at=59c6a479cfeed2eb6541c57e):

> So my game is a spaceship game, sort of like star trek and the like except for you play as a single member of the crew rather than commanding the entire ship. I want to be able to simulate combat and action aboard ships, as well as between ships. This sounds straightforward, until you realize just how much the movement of the ship impacts the crew of it, sci-fi usually handwaves this but it's actually really difficult to reproduce that kind of behavior without using some concept of "recursive worlds" where each ship contains its own mini world, where for all intents and purposes, the ship walls don't move.
It was remarkably difficult to do this in Unity and UE4 without sort of re-inventing the world

> @OvermindDL1 Also mentioned they had use for recursive worlds.

> Which is kind of why I wanted to add this to specs:
> 
> ```rust
> impl Component for World {
>     type Storage = DenseVecStorage<Self>;
> }
> ```

> For three lines you get recursive worlds.

> Basically I needed multiple physics simulations, and the engines would only give me 1
I'd actually kind of given up on having simulated crew activity and was just going to have GUI buttons to let you pick where on the ship your character was sitting, but if we introduce recursive worlds like this I don't have to compromise my vision of my game.
Which is actually really exciting to me.

> Like sure the renderer won't automatically draw my recursive world at the place in the overworld where I want it to be, but I was going to use the recursive world to simulate the gameplay then put those results into entities in the main world, with said entities in the main world not participating in physics themselves.
(One example of problems I faced while doing this in Unity is that if I stuck both the crew and the ship in the main world if a crew member was walking through a narrow hallway towards the front or back of the ship, the rotation of the ship could cause the crewmember to, within a single frame, tunnel out of the walls of the ship.)

---

So first of all, yes you could use `World` as component. But here are the things that would bother me:

* You have to register components on every `World`. That means you'll also all component storages duplicated.
* In case you want to do dispatches with an "inner" world, you cannot access resources of the outer one. That means you had to duplicate that, too.
* I'm sure there are things you'd just want to do with every component, regardless of the ship it's in.

So the most obvious solution that'd come to my mind is, that if you want to only deal with one spaceship, you'd do:

```rust
for (player, ship_id) in (
    &players,
    &ship_ids)
    .join()
    .filter(|&(_, ref id)| id == ShipId(3)) 
{
    // Only deal with players on ship 3
}
```

Is there anything you cannot do with the above pattern?

**Comment by Xaeroxe** [comment]
> Is there anything you cannot do with the above pattern?

The answer to that mostly depends on how flexible the physics systems and components are.  I understand your concerns with this PR though, so perhaps we ought to strive for a more flexible physics system in Amethyst rather than introducing recursion into ECS.  I do want to point out that World could probably do with the Send and Sync impl in this PR though, if for nothing else other than permitting additional flexibility.

**Comment by torkleyy** [comment]
> I do want to point out that World could probably do with the Send and Sync impl in this PR though, if for nothing else other than permitting additional flexibility.

Agreed.

**Comment by torkleyy** [comment]
I think it's a fairly common situation in game physics that you have some sort of "areas".

**Comment by OvermindDL1** [comment]
> You have to register components on every World. That means you'll also all component storages duplicated.

That's fine for me.  In some of my games I registered very different components/systems anyway.  Even on the ones where they were the same I still wanted them separate as, say, doing collision detection between one world and another makes no sense considering their wildly different coordinate systems, and of course that they don't actually touch.  ^.^

> In case you want to do dispatches with an "inner" world, you cannot access resources of the outer one. That means you had to duplicate that, too.

Generally no issue for me either as the asset system handled reuse of resources.

> I'm sure there are things you'd just want to do with every component, regardless of the ship it's in.

Possible, but in many years I've never needed such a use case where duplicating the system in each world was not fine (and in some cases this would be substantially faster as sub-worlds can tick at different/slower rates based on certain conditions).

**Comment by Xaeroxe** [comment]
@torkleyy So I guess some may think it's poor design, others may not.  I don't see the harm in simply permitting recursive worlds, there aren't really any technological limitations to doing this.

**Comment by torkleyy** [comment]
Thanks @Xaeroxe, I think we can just merge this as is.

bors r+

File: src/world.rs [hunk 1]
```diff
@@ -952,6 +952,14 @@ impl World {
     }
 }
 
+unsafe impl Send for World {}
+
+unsafe impl Sync for World {}
```
[the remark this answers is not in the document]
Comment (torkleyy) [reply src/world.rs]: It's correct, trust me.

[the remark this answers is not in the document]
Comment (omni-viral) [reply src/world.rs]: `TrustCell: Sync`?

[the remark this answers is not in the document]
Comment (torkleyy) [reply src/world.rs]: `true`

**Pull request merged**